### DISCUSSION TITLE
Internal Link Validation

### DISCUSSION THREAD

**Discussion by steveoh** [withastro/roadmap#739] [Proposal]
### Body

# Summary

As a content platform it would be an improvement if astro could have build time support for internal link validation

# Background & Motivation

Link validation is important to provide a high quality experience to users of sites built with astro. Spelling mistakes or refactorings of internal pages or assets should not create a mess of 404's and extra work for the creator. Having the DX ability to be confident to merge a pull request that nothing breaks is very important. 

The motivation is to give developers confidence and parity with similar tools to make migrating or choosing astro super easy.
 
# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- View internal broken link assets at build time. Either a warning or error depending on configuation.
- If parsing the AST, building a list of internal vs external assets to enable other features like [notfound](https://github.com/tmcw/notfoundbot) could be a nice side effect.

# Example

https://github.com/HiDeoo/starlight-links-validator/tree/main/packages/starlight-links-validator

https://jekyllrb.com/docs/liquid/tags/#link

```
Failed during stage "building site": Build script returned non-zero exit code: 2
Could not find document "images/my-image.png". Make sure the document exists and the path is correct. in src/content/my-page.md
```

**Comment by y-nk** [comment]
this sounds like `getStaticProps` pages would need to be processed earlier at build phase. it'd be v. nice if we could even expose a cli to list them all (similarly to strapi's `routes:list` command)

**Reply by steveoh** [reply]
it could be nice if external vs internal links were separated and then maybe file types were grouped together

**Comment by steveoh** [comment]
bump

**Comment by steveoh** [comment]
It looks like the astro docs website uses the built sitemap and then parses all the content within it with https://github.com/withastro/docs/blob/main/scripts/lint-linkcheck.ts

Is this something that could be improved and happen during generation rather than after a build and released for all astro users to benefit from?

**Reply by steveoh** [reply]
https://github.com/HiDeoo/starlight-links-validator/issues/28#issuecomment-1914855328

**Comment by stereobooster** [comment]
Related https://github.com/withastro/starlight/discussions/946

**Comment by OliverSpeir** [comment]
This could make a nice integration, but I agree it would be a big win however it gets implemented

**Comment by stereobooster** [comment]
My experimental solution https://astro-digital-garden.stereobooster.com/recipes/detect-broken-links/

**Comment by ghost** [comment]
https://lychee.cli.rs/ could be a great option
